pkg/sfu/sendsidebwe/congestion_detector.go

```diff
@@ -0,0 +1,552 @@
+
+	estimatedAvailableChannelCapacity int64
+	congestionState                   CongestionState
+	congestionStateSwitchedAt         time.Time
+	congestedCTRTrend                 *ccutils.TrendDetector[float64]
+
+	onCongestionStateChange func(congestionState CongestionState, estimatedAvailableChannelCapacity int64)
+}
+
+func NewCongestionDetector(params congestionDetectorParams) *congestionDetector {
+	c := &congestionDetector{
+		params:                            params,
+		packetTracker:                     NewPacketTracker(packetTrackerParams{Logger: params.Logger}),
+		twccFeedback:                      NewTWCCFeedback(twccFeedbackParams{Logger: params.Logger}),
+		wake:                              make(chan struct{}, 1),
+		estimatedAvailableChannelCapacity: 100_000_000,
+	}
+
+	c.feedbackReports.SetMinCapacity(3)
+
+	go c.worker()
+	return c
+}
+
+func (c *congestionDetector) Stop() {
+	c.stop.Once(func() {
+		close(c.wake)
+	})
+}
+
+func (c *congestionDetector) OnCongestionStateChange(f func(congestionState CongestionState, estimatedAvailableChannelCapacity int64)) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.onCongestionStateChange = f
+}
+
+func (c *congestionDetector) GetCongestionState() CongestionState {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.congestionState
+}
+
+func (c *congestionDetector) updateCongestionState(state CongestionState) {
+	c.lock.Lock()
+	c.params.Logger.Infow(
+		"congestion state change",
+		"from", c.congestionState,
+		"to", state,
+		"estimatedAvailableChannelCapacity", c.estimatedAvailableChannelCapacity,
+	)
+
+	prevState := c.congestionState
+	c.congestionState = state
+
+	onCongestionStateChange := c.onCongestionStateChange
+	estimatedAvailableChannelCapacity := c.estimatedAvailableChannelCapacity
+	c.lock.Unlock()
+
+	if onCongestionStateChange != nil {
+		onCongestionStateChange(state, estimatedAvailableChannelCapacity)
+	}
+
+	// when in congested state, monitor changes in captured traffic ratio (CTR)
+	// to ensure allocations are in line with latest estimates, it is possible that
+	// the estimate is incorrect congestion starts and the allocation may be
+	// sub-optimal and not enough to reduce/relieve congestion, by monitoing CTR
+	// on a continuous basis allocations can be adjusted in the direction of
+	// reducing/relieving congestion
+	if state == CongestionStateCongested && prevState != CongestionStateCongested {
+		c.congestedCTRTrend = ccutils.NewTrendDetector[float64](ccutils.TrendDetectorParams{
+			Name:   "ssbwe-estimate",
+			Logger: c.params.Logger,
+			Config: c.params.Config.CongestedCTRTrend,
+		})
+	} else if state != CongestionStateCongested {
+		c.congestedCTRTrend = nil
+	}
+}
+
+func (c *congestionDetector) GetEstimatedAvailableChannelCapacity() int64 {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.estimatedAvailableChannelCapacity
+}
+
+func (c *congestionDetector) HandleRTCP(report *rtcp.TransportLayerCC) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
```

Comment:
have been bitten by "posting to a closed channel" panic enough times :-) . So, tend to lock that as well. But, that is not doing what it is intended here. Will add a check for stop fuse also.

Will `select` not attempt to post if channel is closed? Not able to find good behaviour documentation there.

Reply:
Will merge this for now. Will change it is posting to channel can be done outside the lock. I think there are other places that can change as well if the posting can happen outside.

Reply:
it might make more sense to use `c.stop.Break()` to stop the worker then you don't have to worry about writing to a closed channel. you would never want to close `wake` anyway because as long as `c.stop` isn't broken the worker will infinitely loop selecting the closed wake channel

Reply:
Sounds good. Will make a small PR to just address that.